core/src/Kokkos_Concepts.hpp

```diff
@@ -184,6 +184,84 @@ KOKKOS_IMPL_IS_CONCEPT(work_item_property)
 KOKKOS_IMPL_IS_CONCEPT(hooks_policy)
 }  // namespace Experimental
 
+//
+// is_team_handle
+// - based on
+// https://kokkos.github.io/kokkos-core-wiki/API/core/policies/TeamHandleConcept.html
+//
+template <typename T>
+struct is_team_handle {
+ private:
+  struct TrivialFunctor {
+    void operator()(double &) {}
+  };
+  double lvalueForMethodsNeedingIt_;
+  double *ptrForMethodsNeedingIt_;
+
+  // we use Sum here but any other reducer can be used
+  // since we just want something that meets the ReducerConcept
+  using reduction_to_test_t = ::Kokkos::Sum<double>;
+
+  template <typename U>
+  using m1 = decltype(std::declval<U const &>().team_rank());
+  template <typename U>
+  using m2 = decltype(std::declval<U const &>().team_size());
+  template <typename U>
+  using m3 = decltype(std::declval<U const &>().league_rank());
+  template <typename U>
+  using m4 = decltype(std::declval<U const &>().league_size());
+  template <typename U>
+  using m5 = decltype(std::declval<U const &>().team_shmem());
+  template <typename U>
+  using m6 = decltype(std::declval<U const &>().team_scratch(int{}));
+  template <typename U>
+  using m7 = decltype(std::declval<U const &>().thread_scratch(int{}));
+  template <typename U>
+  using m8 = decltype(std::declval<U const &>().team_barrier());
+  template <typename U>
+  using m9 = decltype(std::declval<U const &>().team_broadcast(
+      lvalueForMethodsNeedingIt_, int{}));
+  template <typename U>
+  using m10 = decltype(std::declval<U const &>().team_broadcast(
+      TrivialFunctor{}, lvalueForMethodsNeedingIt_, int{}));
+
+  template <typename U>
+  using m11 = decltype(std::declval<U const &>().team_reduce(
+      std::declval<reduction_to_test_t>()));
+
+  template <typename U>
+  using m12 = decltype(std::declval<U const &>().team_scan(
+      lvalueForMethodsNeedingIt_, ptrForMethodsNeedingIt_));
+
+ public:
+  static constexpr bool value =
+      is_detected<m1, T>::value && is_detected<m2, T>::value &&
```

Comment:
your's would have succeeded if I return `void` from `team_scratch()` and clearly that would not meet the requirements of a team handle

Reply:
ah yes! good point! in any case, I reworked them all now , i think it is correct now

Reply:
btw, while doing this I noticed that the detection for other concepts (e.g. execution space) is less strict that should be , so maybe at some point it would make sense to rework all those